### PR TITLE
feat(coverage): Add code coverage collection and export

### PR DESCRIPTION
**Pull request by sohil-kshirsagar**
Add code coverage collection to the Tusk Drift CLI. Collects per-test and aggregate code coverage during test replay, with LCOV/JSON export, config-driven activation for CI, and backend upload.

### Coverage activation

Two modes:
- **Config-driven (CI)**: `coverage.enabled: true` in `.tusk/config.yaml` → automatically collects during validation runs on the default branch. Silent (no console output). No CI pipeline changes needed.
- **Flag-driven (local)**: `--show-coverage` displays coverage, `--coverage-output` writes LCOV/JSON.

### Architecture

1. CLI sets `TUSK_COVERAGE=true` + `NODE_V8_COVERAGE=<dir>` env vars
2. After service ready, CLI takes **baseline snapshot** (all coverable lines for denominator)
3. After each test, CLI takes **per-test snapshot** (delta since last reset)
4. Per-test coverage uploaded to backend via `TraceTestCoverageData` on `TraceTestResult`
5. Baseline + startup coverage uploaded via `CoverageBaseline` on `UpdateDriftRunCIStatusRequest`
6. Backend computes aggregate from IN_SUITE tests post-promotion

### Config options

```yaml
coverage:
  enabled: true                    # auto-collect during validation runs
  include: ["backend/src/**"]      # restrict to service code (monorepos)
  exclude: ["**/migrations/**"]    # exclude noise
  strip_path_prefix: "/app"        # Docker: strip container mount point
```

### Key features

- Baseline + per-test snapshot architecture with V8 counter auto-reset
- Branch coverage with union semantics (sum covered, clamp to total)
- LCOV and JSON export formats
- Include/exclude glob filtering with doublestar
- In-suite-only filtering for validation runs (draft tests excluded from output)
- TUI integration: baseline capture, aggregate in service logs, sorted per-test display
- Docker Compose support via `strip_path_prefix`
- Backend upload: per-test coverage piggybacked on test result upload, baseline on CI status update
- Startup coverage included in aggregate (matches industry standard)
- CLI and backend produce identical coverage numbers

### TODOs before merge

- [x] ~~**Blocked on** Use-Tusk/tusk-drift-schemas#47~~ — `go.mod` replace directive removed, updated to v0.1.34
- [x] ~~**Blocked on** Use-Tusk/drift-node-sdk#150 and Use-Tusk/drift-python-sdk#87~~ — SDKs updated, E2E tested with published schemas
- [x] ~~Add LCOV/JSON export tests~~

### Edge cases / gotchas

- `NODE_V8_COVERAGE` is inherited by all child processes. SDK quick-scans to skip non-server files.
- Stale `dist/` artifacts from tsc cause import errors. Use `rm -rf dist` before build.
- Docker paths are container-absolute. `strip_path_prefix` strips the mount point.
- Coverage callback runs BEFORE upload callback so per-test data is available in the proto.
- `GetCoverageBaselineForUpload` merges baseline + per-test records for complete denominator.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -74,3 +74,5 @@ require (
 	golang.org/x/sys v0.39.0 // indirect
 	golang.org/x/text v0.29.0 // indirect
 )
+
+replace github.com/Use-Tusk/tusk-drift-schemas => /Users/sohil/projects/tusk-drift-schemas
```
[the remark this answers is not in the document]
Comment (sohil-kshirsagar) [reply go.mod]: Known release step — the `replace` directive will be removed when `tusk-drift-schemas` is published (documented in PR TODOs). Not a bug.

*— Claude Code*

File: internal/runner/executor.go [hunk 2]
```diff
@@ -473,6 +488,157 @@ func (e *Executor) SetOnTestCompleted(callback func(TestResult, Test)) {
...
+				existing.Lines[line] = count
+			}
+		}
+		// Merge branch data (keep max per line)
+		for line, branchInfo := range fileData.Branches {
+			if existing.Branches == nil {
+				existing.Branches = make(map[string]BranchInfo)
+			}
+			if eb, ok := existing.Branches[line]; !ok || branchInfo.Total > eb.Total {
+				existing.Branches[line] = branchInfo
```
[the remark this answers is not in the document]
Comment (sohil-kshirsagar) [reply internal/runner/executor.go]: Verified correct — when `branchInfo.Total == eb.Total`, the Total stays the same (correct since both report the same file-level total), and `Covered` is still summed+clamped on the subsequent lines. The sum+clamp handles the union correctly regardless of whether Total changed.

*— Claude Code*

**Review by sohankshirsagar — APPROVED** [review]

File: internal/runner/coverage.go [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+
+// CoverageExport is the top-level JSON export structure.
+type CoverageExport struct {
+	Summary   CoverageSummary                        `json:"summary"`
+	Aggregate CoverageSnapshot                       `json:"aggregate"`
+	PerTest   map[string]map[string]CoverageFileDiff `json:"per_test"`
+}
+
+// WriteCoverageJSON writes aggregate + per-test coverage as JSON.
+func WriteCoverageJSON(path string, aggregate CoverageSnapshot, perTest map[string]map[string]CoverageFileDiff, records []CoverageTestRecord) error {
```
Comment (jy-tan) [inline internal/runner/coverage.go]: should `WriteCoverageJSON()`, `printCoverageSummary()`, and `FormatCoverageSummaryLines()` call a shared helper for consistency?

something like `BuildCoverageReportView(rawSession, reportOptions)` that applies suite filtering and include/exclude filtering exactly once.

```go
type CoverageReportView struct {
    Records   []CoverageTestRecord
    Aggregate CoverageSnapshot
    PerTest   map[string]map[string]CoverageFileDiff
    Summary   CoverageSummary
}
```

- `printCoverageSummary()` uses `view.Summary`
- `FormatCoverageSummaryLines()` uses `view.Summary`
- `WriteCoverageJSON()` writes `view.Aggregate`, `view.PerTest`, and `view.Summary`

**Review by jy-tan — COMMENTED** [review]
should `WriteCoverageJSON()`, `printCoverageSummary()`, and `FormatCoverageSummaryLines()` call a shared helper for consistency?

something like `BuildCoverageReportView(rawSession, reportOptions)` that applies suite filtering and include/exclude filtering exactly once.

```
type CoverageReportView struct {
    Records   []CoverageTestRecord
    Aggregate CoverageSnapshot
    PerTest   map[string]map[string]CoverageFileDiff
    Summary   CoverageSummary
}
```
- `printCoverageSummary()` uses `view.Summary`
- `FormatCoverageSummaryLines()` uses `view.Summary`
- `WriteCoverageJSON()` writes `view.Aggregate`, `view.PerTest`, and `view.Summary`

File: internal/runner/coverage.go (hunk 3 above)
Comment (sohil-kshirsagar) [reply internal/runner/coverage.go]: done — added `CoverageReportView` struct with a `BuildCoverageReportView` builder. the filter → merge → pattern → summary chain runs once, and `FormatCoverageSummaryLines` + `ProcessCoverageWithAggregate` both consume the view. `printCoverageSummary` now takes a `CoverageSummary` directly instead of recomputing it.

**Review by jy-tan — APPROVED** [review]
